src/Std/Data/HashSet/Basic.lean

```diff
@@ -0,0 +1,154 @@
+/-
+Copyright (c) 2024 Lean FRO, LLC. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Markus Himmel
+-/
+prelude
+import Std.Data.HashMap.Basic
+
+/-!
+# Hash sets
+
+This file develops the type `Std.Data.HashSet` of dependent hash sets.
+
+The operations `map` and `filterMap` on `Std.Data.HashSet` are defined in the module
+`Std.Data.HashSet.AdditionalOperations`.
+
+Lemmas about the operations on `Std.Data.HashSet` are available in the
+module `Std.Data.HashSet.Lemmas`.
+
+See the module `Std.Data.HashSet.Raw` for a variant of this type which is safe to use in
+nested inductive types.
+-/
+
+set_option linter.missingDocs true
+set_option autoImplicit false
+
+universe u v
+
+variable {α : Type u}
+
+namespace Std
+
+/-- Hash sets. -/
+structure HashSet (α : Type u) [BEq α] [Hashable α] where
+  /-- Internal implementation detail of the hash set. -/
+  inner : HashMap α Unit
+
+namespace HashSet
+
+/-- Creates a new empty hash set. The optional parameter `capacity` can be supplied to presize the
+map so that it can hold the given number of elements without reallocating. It is also possible to
+use the empty collection notations `∅` and `{}` to create an empty hash map with the default
+capacity. -/
+@[inline] def empty [BEq α] [Hashable α] (capacity := 8) : HashSet α :=
+  ⟨HashMap.empty capacity⟩
+
+instance [BEq α] [Hashable α] : EmptyCollection (HashSet α) where
+  emptyCollection := empty
+
+/-- Insert the given element into the set. -/
+@[inline] def insert [BEq α] [Hashable α] (m : HashSet α) (a : α) : HashSet α :=
+  ⟨m.inner.insertIfNew a ()⟩
+
+/-- Equivalent to (but potentially faster than) calling `contains` followed by `insert`. -/
+@[inline] def containsThenInsert [BEq α] [Hashable α] (m : HashSet α) (a : α) : Bool × HashSet α :=
```

Comment:
This is useful when the `BEq` and `Hashable` instances are lawful, but coarser than propositional equality (e.g. terms identified up to alpha-equivalence). Should there be a note to that effect here? Otherwise, readers might wonder what this operation is for.

Reply:
In my experience, it's not so unusual to have this use case of "I want to insert this, and also know whether I have just inserted this for the first time". Rust and Java certainly also support this use case in their `insert`/`add` methods and there were countless occurrences of this in the Java code base I worked on at my previous job. So I don't think we need to say more here.